### ISSUE TITLE
Problems when converting PMTiles to MBTiles in the Python interface

### ISSUE THREAD

**Issue by nokonoko1203** [protomaps/PMTiles#120]
Thank you for defining a great data format.

I am sharing this because I have encountered a problem when converting PMTiles using the Python interface.

After installing the library from PyPI (https://pypi.org/project/pmtiles/), I tried to convert some PMTiles data (https://github.com/amx-project/a-spec) to MBTiles. I got the following exception.

```python
from pmtiles.convert import pmtiles_to_mbtiles


input_file = "a.pmtiles"
output_file = "a.mbtiles"

pmtiles_to_mbtiles(input_file, output_file)
```

```bash
---------------------------------------------------------------------------
ProgrammingError                          Traceback (most recent call last)
Cell In[2], line 7
      5 if os.path.exists(input_file):
      6     os.remove(output_file)
----> 7 pmtiles_to_mbtiles(input_file, output_file)

File ~/miniforge3/envs/pmtiles/lib/python3.11/site-packages/pmtiles/convert.py:129, in pmtiles_to_mbtiles(input, output)
    126         metadata["format"] = "pbf"
    128 for k, v in metadata.items():
--> 129     cursor.execute("INSERT INTO metadata VALUES(?,?)", (k, v))
    131 for zxy, tile_data in all_tiles(source):
    132     flipped_y = (1 << zxy[0]) - 1 - zxy[2]

ProgrammingError: Error binding parameter 2: type 'dict' is not supported
```

Insert the following code at [here](https://github.com/protomaps/PMTiles/blob/28df71c58889f4ac0fd09a9caa95c8797b7877bb/python/pmtiles/convert.py#L128) I inserted the following code.

```python
    for k, v in metadata.items():
        print(f "v_type: {type(v)}")
        print("\n")

        cursor.execute("INSERT INTO metadata VALUES(? ,?)" , (k, v))
```

According to the exception message, v contained a variable of type dict.

```bash
...
v_type: <class 'dict'>


---------------------------------------------------------------------------
ProgrammingError                          Traceback (most recent call last)
Cell In[15], line 32
     29         metadata["format"] = "pbf"
     31 for k, v in metadata.items():
---> 32     cursor.execute("INSERT INTO metadata VALUES(?,?)", (k, v))

ProgrammingError: Error binding parameter 2: type 'dict' is not supported
```

The PMTiles v3 spec states that JSON is to be stored in the metadata, and it seems that MBTiles can also store arbitrary information.
- PMTiles spec: https://github.com/protomaps/PMTiles/blob/main/spec/v3/spec.md
- MBTiles spec: https://github.com/mapbox/mbtiles-spec/blob/master/1.3/spec.md#metadata

Therefore, why not use the sqlite3 library to insert the following code before executing the INSERT query and replace it with a JSON string?

```python
    for k, v in metadata.items():
        if not isinstance(v, str):
            v = json.dumps(v, ensure_ascii=False)

        cursor.execute("INSERT INTO metadata VALUES(? ,?)" , (k, v))
```

I will create a PR.

https://github.com/protomaps/PMTiles/pull/121

**Comment by bdon** [comment]
MBTiles cannot store arbitrary information, only string keys and string values, so any non-string needs to be JSON encoded. If we directly translate between PMTiles JSON metadata and MBTiles we would have JSON-encoded values in the PMTiles JSON (double-encoded), which I want to avoid. The code that goes from MBTiles to PMTiles should special case the key called `json` in the MBTiles table, but does not act on arbitrary other keys.

For converting from PMTiles into sqlite metadata we might need to add special behavior to detect keys called `vector_layers` or `tilestats` and lift those into a `json` row. Does the metadata you are working with use those specific keys?

We might also need some special behavior to stringify numeric values if present in PMTiles (though the SQLite driver might handle that automatically)

**Comment by nokonoko1203** [comment]
> For converting from PMTiles into sqlite metadata we might need to add special behavior to detect keys called vector_layers or tilestats and lift those into a json row. Does the metadata you are working with use those specific keys?

Yes, that is correct.
I am using the `tilestats` key and the `vector_layers` key.

**Comment by bdon** [comment]
In that case, the code should check if the key name is `tilestats` or `vector_layers` and then: parse those JSON values, insert them into a new Dict, stringify, and then insert into SQLite with the key `json`. That would make your result MBTiles conform to the MBTiles spec https://github.com/mapbox/mbtiles-spec/tree/master/1.3 

Unfortunately that logic is a bit messy, but we should have only have to special case it for those two keys. What do you think?

**Comment by nokonoko1203** [comment]
First, the metadata of my PMTiles contains [Vector_layers](https://github.com/mapbox/mbtiles-spec/blob/master/1.3/spec.md#vector_layers) and [Tilestats](https://github.com/mapbox/mbtiles-spec/blob/master/1.3/spec.md#tilestats), which I believe is exactly the nested JSON.
And I think I understand that these PMTiles need to be stored in the TEXT type of sqlite when stored in the MBTiles metadata table.

Also, the python json module seems to convert ints and floats to str as well, as shown below.
(Of course, dict and list are also replaced by string).

```python
Python 3.8.15 | packaged by conda-forge | (default, Jan 26 2023, 10:51:07) 
[Clang 14.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import json
>>> json.dumps(1)
'1'
>>> json.dumps(True)
'true'
>>> json.dumps(1.1111)
'1.1111'
>>> json.dumps({"key": "value"})
'{"key": "value"}'
>>> json.dumps([{"key1": "value1"}, {"key2": "value2"}])
'[{"key1": "value1"}, {"key2": "value2"}]'
```

Therefore, we thought it would be acceptable to pass all values to the dumps method as follows.

```python
    for k, v in metadata.items():
        if not isinstance(v, str):
            v = json.dumps(v, ensure_ascii=False)

        cursor.execute("INSERT INTO metadata VALUES(? ,?)" , (k, v))
```

If my understanding is wrong, I apologize.
Or is there some other reason?

**Comment by bdon** [comment]
Hi @nokonoko1203 , I think your understanding is correct. I have not finished this issue yet: https://github.com/protomaps/PMTiles/issues/111#issuecomment-1386558336

How are you currently creating the PMTiles that you are then trying to convert into MBTiles?

The end goal should be for the PMTiles JSON metadata to have top-level keys, `vector_layers` and `tilestats` instead of having a key called `json`.

If you are using the `go-pmtiles` tool, it is currently using this to avoid json-inside-json: https://github.com/protomaps/go-pmtiles/blob/main/pmtiles/convert.go#L522

Tippecanoe will also create metadata in this format.

However, the python `pmtiles-convert` does not have this logic yet. 

I realize that this should be part of the specification, apologies for the confusion.

**Comment by bdon** [comment]
Thanks, I can look at in the viewer here: https://protomaps.github.io/PMTiles/?url=https%3A%2F%2Fx.optgeo.org%2Fipfs%2FQmTZHWMAnRC5zNiNvdVuTDacThKkj4jKbwsZtKQkAC4R69#map=4.22/35.8/134.38

https://github.com/protomaps/PMTiles/issues/111 Here is the proposed addition to the spec, to make the PMTiles->MBTiles conversion result in the correct metadata, that will need to be added to the Python program.

**Comment by bdon** [comment]
PR now open https://github.com/protomaps/PMTiles/pull/204

that lifts `vector_layers` and `tilestats` into the `json` key when converting to MBTiles.

**Comment by bdon** [comment]
Resolved in [python pmtiles v3.2.0](https://pypi.org/project/pmtiles/)
